File: lib/scala/generic/src/main/scala/org/paradicms/lib/generic/models/Person.scala
package org.paradicms.lib.generic.models

import io.lemonlabs.uri.Uri
import org.paradicms.lib.generic.models.domain.{DomainModel, DomainModelCompanion}

final case class Person(
                         familyName: Option[String],
                         givenName: Option[String],
                         name: Option[String],
                         sortName: Option[String],
                         uri: Uri
                       ) extends DomainModel

object Person extends DomainModelCompanion {
  def apply(resource: ResourceWrapper): Person =
    Person(
      familyName=resource.foaf.familyName(),
      givenName=resource.foaf.givenName(),
      name=resource.foaf.name(),
      sortName=resource.contact.sortName(),
      uri=resource.uri
    )
}

File: etl/paradicms_etl/models/person.py
from dataclasses import dataclass
from typing import Optional

from rdflib import Graph, Literal
from rdflib.namespace import FOAF, RDF
from rdflib.resource import Resource

from paradicms_etl.models._named_model import _NamedModel
from paradicms_etl.namespace import CMS, CONTACT
from paradicms_etl.utils.rdf_resource_wrapper import RdfResourceWrapper


@dataclass(frozen=True)
class Person(_NamedModel):
    name: str
    family_name: Optional[str] = None
    given_name: Optional[str] = None
    sort_name: Optional[str] = None

    def to_rdf(self, *, graph: Graph) -> Resource:
        resource = _NamedModel.to_rdf(self, graph=graph)
        if self.family_name is not None:
            resource.add(FOAF.familyName, Literal(self.family_name))
        if self.given_name is not None:
            resource.add(FOAF.givenName, Literal(self.given_name))
        resource.add(FOAF.name, Literal(self.name))
        if self.sort_name is not None:
            resource.add(CONTACT.sortName, Literal(self.sort_name))
        return resource
